[5.5] [WIP] Transformable Responses

Aferz opened the pull request:
### Why ?

The purpose of this pull request is to give **Laravel** better flexibility when responding to API requests.

I have created a new `TransformableResponse` class that allows to define a few rulesets for transform the response in multiple forms that will be passed to the client.

Its design is highly inspired in `FormRequests` classes. If we can define `Requests` individually, _why don't extend this behaviour to responses_?

### What is included in this PR ?

First of all, I suggest to store our `Response` classes inside `app/Http/Responses` folder(A nice artisan command will be included out of the box). Once said this ...

#### Class structure

```php
class GetTestsResponse extends TransformableResponse
{
    /**
     * Set visibility rules to apply to current response.
     *
     * @return array
     */
    public function visibilityRules()
    {
        return [];
    }
    
   /**
     * Set casting rules to apply to current response.
     *
     * @return array
     */
    public function castingRules()
    {
        return [];
    }

    /**
     * Set mutation rules to apply to current response.
     *
     * @return array
     */
    public function mutationRules()
    {
        return [];
    }
    
    /**
     * Set renaming rules to apply to current response.
     *
     * @return array
     */
    public function renamingRules()
    {
        return [];
    }
}
```

#### Dummy Data

We will use this dummy data for demo purposes:

```php
[
  'id' => '1',
  'name' => 'Name #1',
  'surname' => 'Surname #1',
  'posts' => [
      [
          'id' => '1',
          'title' => 'Post #1',
          'comments' => [
              [
                  'id' => '1',
                  'title' => 'Comment #1-1'
              ]
          ]
      ],
      [
          'id' => '2',
          'title' => 'Post #2',
          'comments' => [
              [
                  'id' => '1',
                  'title' => 'Comment #2-1'
              ]
          ]
      ]
  ]
]
```


#### Defining visibility rules

This ruleset define what fields will be present once the response is emitted. It's the **first** ruleset to be executed.

```php
   /**
     * Set visibility rules to apply to current response.
     *
     * @return array
     */
    public function visibilityRules()
    {
        return [
            'id' => true,
            'name' => true,
            'posts.*.id' => true,
            'posts.*.comments.*.id' => true
        ];
    }
```

This will be the resultant response:

```json
{
   "id":"1",
   "name":"Name #1",
   "posts":[
      {
         "id":"1",
         "comments":[
            {
               "id":"1"
            }
         ]
      },
      {
         "id":"2",
         "comments":[
            {
               "id":"1"
            }
         ]
      }
   ]
}
```

##### How does this work ?

As you can see in the rules section you define which attributes should be visible and which attributes should be hidden. So:

1. Fields checked as **true** will be evaluated first. If none, **all** will be visible by **default**.
2. Fields checked as **false** will be hidden using the data passed from the previous process.

Easy.

#### Defining casting rules

This ruleset define which fields should be casted. Valid casting values are: `int, integer, real, float, double, string, bool, boolean`. It's the **second** ruleset to be executed.

```php
    /**
     * Set casting rules to apply to current response.
     *
     * @return array
     */
    public function castingRules()
    {
        return [
            'id' => 'int',
            'posts.*.id' => 'int',
            'posts.*.comments.*.id' => 'int'
        ];
    }
```

This will be the resultant response (chained from first ruleset):

```json
{
   "id":1,
   "name":"Name #1",
   "posts":[
      {
         "id":1,
         "comments":[
            {
               "id":1
            }
         ]
      },
      {
         "id":2,
         "comments":[
            {
               "id":1
            }
         ]
      }
   ]
}
```

#### Defining mutation rules

This ruleset define how the data should be mutated. You can define your own mutators in the own class, just prefix the name of your method with the word `mutator`. If you want to reuse mutators you always can inherit from a parent class 👍 . It's the **third** ruleset to be executed.

```php
    /**
     * Set mutation rules to apply to current response.
     *
     * @return array
     */
    public function mutationRules()
    {
        return [
            'id' => 'double|uppercase',
            'name' => 'uppercase',
            'posts.*.id' => 'double',
            'posts.*.comments.*.id' => 'double'
        ];
    }

    public function mutatorDouble($value)
    {
        return $value * 2;
    }
    
    public function mutatorUppercase($value)
    {
        return strtoupper($value);
    }
```

This will be the resultant response (chained from first and second ruleset):

```json
{
   "id":"2",
   "name":"NAME #1",
   "posts":[
      {
         "id":2,
         "comments":[
            {
               "id":2
            }
         ]
      },
      {
         "id":4,
         "comments":[
            {
               "id":2
            }
         ]
      }
   ]
}
```

#### Defining renaming rules

This ruleset define how the keys of your data should be renamed. It's the **forth** ruleset to be executed.

```php
    /**
     * Set renaming rules to apply to current response.
     *
     * @return array
     */
    public function renamingRules()
    {
        return [
            'id' => 'user_id',
            'posts.*.id' => 'post_id',
            'posts.*.comments.*.id' => 'comment_id'
        ];
    }
```

This will be the resultant response (chained from first, second and third ruleset):

```json
{
   "user_id":"2",
   "name":"NAME #1",
   "posts":[
      {
         "post_id":2,
         "comments":[
            {
               "comment_id":2
            }
         ]
      },
      {
         "post_id":4,
         "comments":[
            {
               "comment_id":2
            }
         ]
      }
   ]
}
```

#### How this feature will be used ?

Simple. Look at this controller method:

```php
    public function index()
    {
        $users = User::all(); // Our dummy data
        
        return new GetTestsResponse($users, 200, []); // status and headers are optional
    }
```

### Why is a WIP ?

There are a few things to be covered before merge this PR, but first I want to check if this is useful for the Laravel community.

**TODO**:
* `php artisan make:response MyCustomResponse` command.
* Tests to cover this PR.

Any idea ? Any feedback ? Let me know what you all think!

Thanks !

Aferz committed:
Added transformable Response

Aferz committed:
Applied CI Styles Fix

feryardiant commented:
Woa, really nice to see someone come to this idea.

I've done something similar before, using @thephpleague/fractal on Laravel 5.4, the implementation was pretty strike forward, nothing's comlicated & it support Paginators well.

**IMO** I'm go with fractal.

Aferz commented:
Yeah, before make this PR o reviewed Fractal, but IMO I think that It doesn't really fit with a nice Laravel approach. For me It was too verbose.

I thought that filter through data needed to be easier, that's why I used validation rules style w\ custom atomic clases to represent a specific response for a request instead of injecting multiple classes.

Thanks for your feedback!!

konigbach commented:
Nice and elegant approach! Fractal is more confuse and verbose and this gives us the simplicity to transform the data the way it should be. Great job!

ggcasuso commented:
Definitely more useful than Fractal! Good work!

devcircus commented:
It may be overkill, but did you consider adding the ability to mutate the data? We can already define mutators and casts, but it would be nice to be able to define all of this in one custom response class. With this, I would drop Fractal completely. However, you probably don't want to add a bunch of unneeded complexity.

Either way, nice work! If not accepted, I'll probably fork and use in some of my api projects anyway.

Aferz commented:
@devcircus What do you mean with mutate the data ?

Do you mean mutate with something like `posts.*.commentsCount + 1` ? If yes, I won't because in my opinion, response object is not responsable for mutate the data this way.

If you mean mutate the schema (replace `A` key for `B` key when responding), yes, that's something that I have in mind if this PR goes forward.

I'm glad you like it. Thanks!

JairoRC commented:
+1! very useful

devcircus commented:
By mutate, I mean simply altering the presentation, but yes, (A) changing the reference(key) and/or (B) changing how the data is presented, similar to how Fractal does it.

This is something that can already be handled by presenters, accessors, etc. I just thought, notwithstanding SRP, that it would be nice to be able to fully define the response inside a custom response class.

deleugpn commented:
Maybe a decorator outside of core would be better for that?

Aferz commented:
Ok, now I see what you mean.

I don't really mind to try to figure out how we could encapsulate that presentation logic if that does not reject this PR (because that's not the main purpose of this feature).

Let's see what @taylorotwell or @GrahamCampbell thinks about this.

Aferz committed:
WIP

Aferz committed:
Applied CI Styles Fix

GrahamCampbell commented on src/Illuminate/Http/TransformableResponse.php:
```diff
@@ -27,102 +27,104 @@ public function __construct($data = null, $status = 200, $headers = [], $options
      * Transforms the response data.
      *
-     * @param  array  $data
+     * @param  array $data
```
please revert

GrahamCampbell commented on src/Illuminate/Http/TransformableResponse.php:
```diff
@@ -0,0 +1,333 @@
+     * @param  int    $status
+     * @param  array  $headers
+     * @param  int    $options
+     */
```
missing return void annotation

Aferz replied:
I copied the constructor from JsonResponse, should I apply equally?

GrahamCampbell replied:
Looks like a mistake there then. Laravel always includes return annotations, on all methods, without exception.

GrahamCampbell commented on src/Illuminate/Http/TransformableResponse.php:
```diff
@@ -0,0 +1,333 @@
+    /**
+     * Transforms the response data.
+     *
+     * @param  array $data
```
cs

GrahamCampbell commented on src/Illuminate/Http/TransformableResponse.php:
```diff
@@ -0,0 +1,333 @@
+     *
+     * Example: posts.*.comments.*.title => posts.0.comments.0.title
+     *
+     * @param  array $data
```
cs

Aferz commented:
Hi guys,

I've updated the main message to reflect some updates I made in this package:

* Casting Rules
* Mutation Rules
* Renaming Rules

Thanks

GrahamCampbell commented on src/Illuminate/Http/TransformableResponse.php:
```diff
@@ -0,0 +1,333 @@
+     * Sanitize gathered array rules removing those that don't appear into
+     * orginal rules.
+     *
+     * @param  string $rule
```
cs, etc

GrahamCampbell commented on src/Illuminate/Http/TransformableResponse.php:
```diff
@@ -0,0 +1,333 @@
+
+    /**
+     * Sanitize gathered array rules removing those that don't appear into
+     * orginal rules.
```
please shorten to 80 chars including 5 spaces, star, and space

Aferz replied:
80 chars length only applies to DocBlocks ?

GrahamCampbell replied:
It's something we apply (ish) to the short descriptions to stop them getting too long. One rule that is strict is that they cannot overflow onto the next line.

GrahamCampbell commented on src/Illuminate/Http/TransformableResponse.php:
```diff
@@ -0,0 +1,333 @@
+    {
+        switch ($this->getCastType($attribute)) {
+            case 'int':
+            case 'integer':
```
lots of code duplication here with eloquent

Aferz replied:
Yep, thats something I wanted to talk here.

I didn't want to refactor Eloquent to use the switch that it uses to define the casting. What should I do here ?

kavinsky commented:
👍 I like the second approach, but casting and mutating should be unified.

GrahamCampbell commented on src/Illuminate/Http/TransformableResponse.php:
```diff
@@ -0,0 +1,339 @@
+     * Constructor.
+     *
+     * @param  mixed  $data
+     * @param  int    $status
```
Please use exactly 2 spaces before the dollar for param annotations on new code, rather than aligning.

GrahamCampbell commented on src/Illuminate/Http/TransformableResponse.php:
```diff
@@ -0,0 +1,339 @@
+    protected function getCastType($attribute)
+    {
+        if (! $castings = $this->castingRules()) {
+            return null;
```
just `return;` gives the same semantics

GrahamCampbell commented on src/Illuminate/Http/TransformableResponse.php:
```diff
@@ -0,0 +1,339 @@
+     * Get casting rule for given attribute.
+     *
+     * @param  string  $attribute
+     * @return bool
```
`bool|null`

GrahamCampbell commented on src/Illuminate/Http/TransformableResponse.php:
```diff
@@ -0,0 +1,339 @@
+
+    /**
+     * Sanitize gathered array rules removing those that
+     * don't appear into orginal rules.
```
these descriptions cannot span more than one line.

GrahamCampbell commented on src/Illuminate/Http/TransformableResponse.php:
```diff
@@ -0,0 +1,339 @@
+
+    /**
+     * Resolve array rules generating a new rule for every
+     * item in the array specified with the '*' symbol.
```
same

fernandobandeira commented:
I think you should remove abstract from those methods since most of the times we won't use the 4 methods on our child class... Making them optional would be better IMO.

Also this seems pretty good so far... :+1:

lucasmichot commented on src/Illuminate/Http/TransformableResponse.php:
```diff
@@ -0,0 +1,335 @@
+     */
+    public function __construct($data = null, $status = 200, $headers = [], $options = 0)
+    {
+        if (is_array($data)) {
```
What if data implements `\Illuminate\Contracts\Support\Jsonable`?

lucasmichot replied:
I would also allow any object to implement a `\Illuminate\Contracts\Support\Transformable` interface, so that the object itself can always control the way it's transformed.

Aferz replied:
How we should treat `\Illuminate\Contracts\Support\Jsonable` ?? I mean, I have no way to determine how to transform the data if I can get it. Any suggestion ?

I like the idea of `\Illuminate\Contracts\Support\Transformable`, but how do we should treat a `Collection` of `Transformables` ? Should `Transformable` prevail against `Arrayable` ?

lucasmichot replied:
> How we should treat \Illuminate\Contracts\Support\Jsonable ??
Have a look at `\Illuminate\Routing\Router::prepareResponse`

> How do we should treat a Collection of Transformables
If data is a collection, map all its items

> Should Transformable prevail against Arrayable
I believe so

Aferz commented:
It seems reasonable. I will modify It.

Thanks

lucasmichot commented on src/Illuminate/Http/TransformableResponse.php:
```diff
@@ -0,0 +1,335 @@
+
+        return array_reduce(array_keys($applicableRules),
+            function ($transformedData, $rule) {
+                if (Arr::has($transformedData, $rule)) {
```
`\Illuminate\Support\Arr::forget` always make sure that the array key exist, so using `\Illuminate\Support\Arr::has` is useless.

lucasmichot commented on src/Illuminate/Http/TransformableResponse.php:
```diff
@@ -0,0 +1,335 @@
+     * @param  mixed  $value
+     * @param  string  $mutators
+     * @return mixed
+     */
```
Missing `@throws \Exception`

lucasmichot commented on src/Illuminate/Http/TransformableResponse.php:
```diff
@@ -0,0 +1,335 @@
+            if (! method_exists($this, $method)) {
+                $classname = static::class;
+
+                throw new \Exception("There is no mutator [$method] declared in [$classname].");
```
Please import `\Exception`

lucasmichot commented on src/Illuminate/Http/TransformableResponse.php:
```diff
@@ -0,0 +1,335 @@
+
+        return array_reduce($gatheredRules,
+            function ($validRules, $rule) use ($pattern, $valueForValidOnes) {
+                preg_match($pattern, $rule, $matches);
```
This is enough as we only want to use the `$subject` argument.

```php
if (preg_match($pattern, $rule)) {
    $validRules[$rule] = $valueForValidOnes;
}
```

lucasmichot commented on src/Illuminate/Http/TransformableResponse.php:
```diff
@@ -0,0 +1,335 @@
+    }
+
+    /**
+     * Sanitize rules removing those that don't appear into orginal rules.
```
Typo: `original `

lucasmichot commented on src/Illuminate/Http/TransformableResponse.php:
```diff
@@ -0,0 +1,335 @@
+    }
+
+    /**
+     * Apply visbility rules to given data.
```
Typo: `visibility`

Aferz committed:
Fix typos and useless codes.

lucasmichot commented:
@Aferz - how do you handle `$original` variable ?
If I pass a a Collection as data in my custom Response, I want to be sure to actually retrieve it

So `$this->getOriginal()` should return my `Collection`
And `$this->getData()` should return my transformed content...

Aferz commented:
@lucasmichot

This:
```php
// Constructor stuff ...

parent::__construct($data, $status, $headers, $options);

$this->original = $data;
```
should do the trick.

* `$this->getOriginaContentl()` will return given Collection or data.
* `$this->getData()` will return transformed content.

I will update this PR with your suggestions tomorrow.

Thanks for the feedback!

jaketoolson commented:
If this gets approved yet #14357 (simple Transformer) PR was not... that would be a shame.

Aferz committed:
Added Transformable Contract. Improved renaming feature. Improved Collection handling.

Aferz commented:
I've updated this PR with following changes:

- Now you can access original data passed to custom response with `getOriginalData()` parent method.
- Removed abstract methods to cherry pick what ruleset should be applied to custom response.
- Added `Illuminate\Support\Contracts\Transformable` to allow object to define how it should be mapped.
- If data is a `Illuminate\Support\Collection` class, its items are mapped checking (In order):
     - `$item instanceof Transformable`
     - or `$item instanceof Arrayable`
     - or returning `$item` itself.
- Improved renaming feature.
- `Paginator`/`LengthAwarePaginator` works right out the box


TODO:
- Command for generate custom responses.
- Tests.

GrahamCampbell commented on src/Illuminate/Http/TransformableResponse.php:
```diff
@@ -0,0 +1,382 @@
+    /**
+     * Handle how a collection will be transformed.
+     *
+     * @param  Collection  $data
```
fqnc required please

Aferz replied:
Oops! Fixed.

Aferz committed:
Fix DockBlock Full Qualified Domain Name.

GrahamCampbell commented on src/Illuminate/Http/TransformableResponse.php:
```diff
@@ -0,0 +1,382 @@
+     *
+     * @param  string  $mutators
+     * @param  mixed  $value
+     * @throws Exception
```
Please follow our cs regarding throws docs. :)

Aferz replied:
Sorry but what do you eaxctly mean ??

GrahamCampbell replied:
https://github.com/laravel/framework/blob/7212b1e9620c36bf806e444f6931cf5f379c68ff/src/Illuminate/Auth/Access/HandlesAuthorization.php#L24

GrahamCampbell commented:
I still wonder if this whole thing would be best suited as a laravel package, rather than shipping in the core.

Aferz committed:
Fixed cs @throws in DocBlock.

Aferz commented:
@GrahamCampbell - That something that we all should talk about, but I'm gonna give my 2 cents.

So far, Laravel offers support/utilities for build an API (Auth & JWT, separated API routes file, Throttling, etc...) but one of the most difficult things is how you display your data and what you show and what you don't for your multiple requests. Yes, you have `$hidden` attribute it your models, but when you try to display, for example, different model data depending on the Request, you don't have enough with a single attribute. That could be the main reason why there are several pull requests about transforming data.

This could be a package? Of course. But IMO I think Laravel should provide 'something' to filter your response before sending to a client (There is no a easy/clean way so far to make this). This way you gain full control from the way your request is validated to how your response is displayed.

So far, Laravel treats every Request as unique (If you want, of course), but not the Response. Conceptually, that's something that could be improved to gain granular control about your code. And that's what this PR wants to achieve.

Anyways, this is my opinion. I would like to know what you all think, if you share my concept about 'unique Response' and if this PR should be accepted or rejected.

I would like to know what @taylorotwell thinks about this to know if this worth it.

VaughanDev commented:
Well :/...

Ok then.

Aferz commented:
Mmmh ok then.

joshmanders commented:
@Aferz 
> There is no a easy/clean way so far to make this

That's not true at all.. Whatever you return from your controller is formatted as a response. You can return anything, and easily intercept and format it however you like. I'm doing it right now. I return a class that extends a certain class, and I typecheck and format. It's pretty easy AND straight forward.

Aferz commented:
Hi @taylorotwell , sorry for bothering you again.

5.5 release is near so... Will this PR take in consideration for 5.5 or higher versions??.

Just for keep updating this PR with useful stuff and improvements or think of it as a possible package.

Thanks

fernandobandeira commented:
@Aferz take a look on #19449
